src/object_store/src/object/opendal_engine.rs

```diff
@@ -0,0 +1,433 @@
+        Self {
+            op,
+            engine_type: EngineType::Hdfs,
+        }
+    }
+
+    /// create opendal memory engine, used for unit tests.
+    pub fn new_memory_engine() -> Self {
+        // Create fs backend builder.
+        let mut builder = memory::Builder::default();
+
+        let op: Operator = Operator::new(builder.build().unwrap());
+        Self {
+            op,
+            engine_type: EngineType::Memory,
+        }
+    }
+
+    async fn get_object<R, F>(&self, path: &str, f: F) -> ObjectResult<R>
+    where
+        F: Fn(&Bytes) -> R,
+    {
+        Some(&Bytes::from(self.op.object(path).read().await?))
+            .ok_or_else(|| ObjectError::internal(format!("no object at path '{}'", path)))
+            .map(f)
+    }
+}
+
+#[async_trait::async_trait]
+impl ObjectStore for OpendalObjectStore {
+    fn get_object_prefix(&self, _obj_id: u64) -> String {
+        String::default()
+    }
+
+    async fn upload(&self, path: &str, obj: Bytes) -> ObjectResult<()> {
+        if obj.is_empty() {
+            Err(ObjectError::internal("upload empty object"))
+        } else {
+            self.op.object(path).write(obj).await?;
+            Ok(())
+        }
+    }
+
+    fn streaming_upload(&self, path: &str) -> ObjectResult<BoxedStreamingUploader> {
+        match self.engine_type {
+            EngineType::Memory => Ok(Box::new(OpendalMemoryStreamingUploader::new(
+                self.op.clone(),
+                path.to_string(),
+            ))),
+            EngineType::Hdfs => Ok(Box::new(HdfsStreamingUploader::new(
+                self.op.clone(),
+                path.to_string(),
+            ))),
+        }
+    }
+
+    async fn read(&self, path: &str, block: Option<BlockLocation>) -> ObjectResult<Bytes> {
+        let res = Bytes::from(self.op.object(path).read().await?);
+        match block {
+            Some(block) => find_block(&res, block),
+            None => Ok(res),
+        }
+    }
+
+    async fn readv(&self, path: &str, block_locs: &[BlockLocation]) -> ObjectResult<Vec<Bytes>> {
+        let futures = block_locs
+            .iter()
+            .map(|block_loc| self.read(path, Some(*block_loc)))
+            .collect_vec();
+        try_join_all(futures).await
+    }
+
+    /// Returns a stream reading the object specified in `path`. If given, the stream starts at the
+    /// byte with index `start_pos` (0-based). As far as possible, the stream only loads the amount
+    /// of data into memory that is read from the stream.
+    async fn streaming_read(
+        &self,
+        path: &str,
+        start_pos: Option<usize>,
+    ) -> ObjectResult<Box<dyn AsyncRead + Unpin + Send + Sync>> {
+        fail_point!("opendal_streaming_read_err", |_| Err(
+            ObjectError::internal("opendal streaming read error")
+        ));
+
+        let bytes = match start_pos {
+            Some(strat_position) => {
+                self.get_object(path, |obj| {
```

Comment:
Same here. Reading from beginning when `start_pos` is provided is a waste of work.